### ISSUE TITLE
Create a polygon of the real data extents of for 3D surveys

### ISSUE THREAD

**Issue by EvanBianco** [agile-geoscience/seisplot#27]
Don't need to show a timeslice on the basemap necessarily, but it would be cool to plot a polygon enclosing the actual data extents relative to the inline and crossline rectangular range.

![image](https://cloud.githubusercontent.com/assets/1699959/16700904/f0dddc36-4532-11e6-9c00-c9d55246e358.png)

**Comment by mycarta** [comment]
This is a possible way with scikit-image, using a time slice:
1) convert to a binary image with appropriate threshold so that no-data is background and data is foreground
2) fill in holes in the foreground that may be caused by areas in the data having had intensity (amplitude) very close to the no-data
3) erode using a very small structuring element
4) your polygon should be the difference between eroded image and non eroded images

Comments: 
- you may have to use more sophysticated morphological operations for (2)
- there should be a perimeter option in skimage regionprops that achieves the same as (3)+(4)

**Comment by kwinkunks** [comment]
I think the computation would be easier with a bit of trace-by-trace processing. Then we can get rid of the zero-crossings. This should work, even for 8-bit data... 

https://github.com/agile-geoscience/seisplot/blob/master/Polygon_around_data_area.ipynb

**Comment by djchild** [comment]
There are algorithms to compute the convex hull of a set of points.

**Comment by mycarta** [comment]
@djchild: seismic surveys are concave, have you manage to make the convex hull work with them? I'd be curious, I have been looking for something like that for some time. The best I could do myself was with the workflow I describe above. 
PS I just bumped into this promising approach with Shapely today: http://blog.thehumangeo.com/2014/05/12/drawing-boundaries-in-python/

**Comment by djchild** [comment]
Good point. It has been over a decade since I worked on that and I don't
recall the details. I do remember that you can speed things up a lot by
discarding most of the traces.

Your suggestion of using Delaunay Triangulation is interesting.
Transforming a gnarly geophysics problem into something that can be solved
with a well known computer science algorithm is often a good approach.

There might be some helpful buzzwords here
http://gis.stackexchange.com/questions/1200/what-are-definition-algorithms-and-practical-solutions-for-concave-hull

**Comment by kwinkunks** [comment]
I improved the contouring method — extracting the contours as `shapely` polygons. So now I get this, which I think is pretty good...

![image](https://cloud.githubusercontent.com/assets/1692372/16844251/c934d482-49ba-11e6-8cc4-b681596477e3.png)

It would be easy to test the Delauney hack from here too, but it looks a bit imprecise to me, and I wonder how it would adapt to holes.

**Comment by EvanBianco** [comment]
I just noticed that Mahotas has had some updates and some improved documentation. There is are a few functions to draw polygons on images. One of them computes the convex hull using the Graham Scan https://en.wikipedia.org/wiki/Graham_scan...See, http://mahotas.readthedocs.io/en/latest/polygon.html

**Comment by mycarta** [comment]
So cool that the Wikipedia link has pseudocode. But, do we NOT want concavities in this case?

**Comment by EvanBianco** [comment]
@mycarta, yes I think you're right.

**Comment by djchild** [comment]
Does this help?
https://medium.com/@agafonkin/a-dive-into-spatial-search-algorithms-ebd0c5e39d2a

He builds the convex hull first and then adjusts it.

**Comment by EvanBianco** [comment]
Concaveman looks cool. I'll be checking this out. Thanks @djchild!

**Comment by kwinkunks** [comment]
Awesome article!

As for Concaveman, I think the problem with a hull is that it won't support
holes in the survey.

**Comment by EvanBianco** [comment]
Related (perhaps it is a new feature?), transform the time slices into proper orientation relative to a given coordinate system. An example in matplotlib is here, https://matplotlib.org/2.0.0/examples/api/demo_affine_image.html

**Comment by mycarta** [comment]
If the amplitude value for the background (null traces, holes) is known, I would go back again to suggest a scikit-image workflow. Let's say  the value is zero; then a logical cutoff can be applied. Something like:

```
bg_range= numpy.logical_and(data< -10^-4,  data > -10^-4)
data[bg_range] = 0
data[numpy.logical_not(bg_range)]= 1
```

after which you should have 1s for live data and 0s elsewhere, including holes. Then you should get a `contour` for each hole plus the boundary. And contour can be accessed individually, sorted, etcetera. I have some examples [here](https://mycarta.wordpress.com/2017/09/18/computer-vision-in-geoscience-recover-seismic-data-from-images-part-2/) but I am also happy to hack if you send a slice or a few my way. This is my recreational drug of choice.

**Comment by kwinkunks** [comment]
@evan Moved to #39

@mycarta I think we solve the outline issue, no? See above. Just needs implementing.

**Comment by mycarta** [comment]
> @evan Moved to #39
> 
> @mycarta I think we solved the outline issue, no? See above. Just needs implementing.

@kwinkunks my bad, it did look like from your last comment that the holes were still a problem, but re-reading again I realize it is about Concaveman
